source/common/router/router.cc

```diff
@@ -846,11 +979,12 @@ void Filter::onUpstreamData(Buffer::Instance& data, UpstreamRequest& upstream_re
     onUpstreamComplete(upstream_request);
   }
 
-  callbacks_->encodeData(data, end_stream);
+  if (final_upstream_request_ == &upstream_request) {
```

Comment:
Looking at this again, I don't think we need this block because when we set `final_upstream_request_` we reset all other upstreams which also removes the callbacks. We might not need `final_upstream_request_` at all actually for that reason